File: packages/llmix/benchmarks/bench_import_time.py
"""
LLMix Python import time benchmark.

Measures how long `import llmix` takes. Target: < 100ms when no
provider SDKs are installed (core-only path).

Run:
    uv run python tests/python/bench_import_time.py
"""

from __future__ import annotations

import sys
import time


def bench_import(iterations: int = 5) -> None:
    """Benchmark ``import llmix`` by forcing a fresh import each run."""
    times: list[float] = []
    for i in range(iterations):
        # Remove cached llmix modules so re-import is measured
        to_remove = [key for key in sys.modules if key == "llmix" or key.startswith("llmix.")]
        for key in to_remove:
            del sys.modules[key]

        start = time.perf_counter()
        import llmix  # noqa: F401
        elapsed_ms = (time.perf_counter() - start) * 1000
        times.append(elapsed_ms)
        print(f"  Run {i + 1}: {elapsed_ms:.1f}ms")

    avg = sum(times) / len(times)
    best = min(times)
    worst = max(times)
    print(f"\nResults ({iterations} iterations):")
    print(f"  Average: {avg:.1f}ms")
    print(f"  Best:    {best:.1f}ms")
    print(f"  Worst:   {worst:.1f}ms")
    print(f"  Target:  < 100ms")
    if avg < 100:
        print("  Status:  PASS")
    else:
        print("  Status:  ABOVE TARGET (may include provider SDK imports)")


if __name__ == "__main__":
    print("LLMix Python import time benchmark\n")
    bench_import()

File: packages/llmix/benchmarks/bench-import-time.ts
/**
 * LLMix TypeScript import time benchmark.
 *
 * Measures how long `import("../typescript/src/index.ts")` takes.
 * Target: < 50ms.
 *
 * Run:
 *   bun run packages/llmix/benchmarks/bench-import-time.ts
 */

export {};

const ITERATIONS = 5;

async function benchImport(): Promise<void> {
  const times: number[] = [];

  for (let i = 0; i < ITERATIONS; i++) {
    // Bun caches modules, so after the first import subsequent ones will
    // hit cache. We measure both to show cold vs warm performance.
    const start = performance.now();
    await import("../typescript/src/index");
    const elapsed = performance.now() - start;
    times.push(elapsed);
    console.log(`  Run ${i + 1}: ${elapsed.toFixed(1)}ms`);
  }

  const avg = times.reduce((a, b) => a + b, 0) / times.length;
  const best = Math.min(...times);
  const worst = Math.max(...times);

  console.log(`\nResults (${ITERATIONS} iterations):`);
  console.log(`  Average: ${avg.toFixed(1)}ms`);
  console.log(`  Best:    ${best.toFixed(1)}ms`);
  console.log(`  Worst:   ${worst.toFixed(1)}ms`);
  console.log(`  Target:  < 50ms`);
  console.log(`  Status:  ${avg < 50 ? "PASS" : "ABOVE TARGET (may include provider SDK imports)"}`);
}

console.log("LLMix TypeScript import time benchmark\n");
await benchImport();
